modeshape-jcr/src/main/java/org/modeshape/jcr/value/basic/ModeShapeDateTime.java

```diff
@@ -0,0 +1,236 @@
+        this.millisInUtc = utcMillis(instance);
+    }
+
+    public ModeShapeDateTime( long milliseconds,
+                              String timeZoneId ) {
+        this.instance = ZonedDateTime.ofInstant(Instant.ofEpochMilli(milliseconds), ZoneId.of(timeZoneId));
+        this.millisInUtc = utcMillis(instance);
+    }
+
+    public ModeShapeDateTime( int year,
+                              int monthOfYear,
+                              int dayOfMonth,
+                              int hourOfDay,
+                              int minuteOfHour,
+                              int secondOfMinute,
+                              int millisecondsOfSecond ) {
+        int nano = (int) TimeUnit.NANOSECONDS.convert(millisecondsOfSecond, TimeUnit.MILLISECONDS);
+        this.instance = ZonedDateTime.of(year, monthOfYear, dayOfMonth, hourOfDay, minuteOfHour, secondOfMinute, nano, UTC);
+        this.millisInUtc = instance.toInstant().toEpochMilli();
+    }
+
+    public ModeShapeDateTime( java.util.Date jdkDate ) {
+        this.instance = ZonedDateTime.ofInstant(jdkDate.toInstant(), TimeZone.getDefault().toZoneId());
+        this.millisInUtc = utcMillis(this.instance);
+    }
+
+    public ModeShapeDateTime( java.util.Calendar jdkCalendar ) {
+        this.instance = ZonedDateTime.ofInstant(jdkCalendar.toInstant(), jdkCalendar.getTimeZone().toZoneId());
+        this.millisInUtc = utcMillis(instance);
+    }
+    
+    protected ModeShapeDateTime( ZonedDateTime zonedDateTime ) {
+        this.instance = zonedDateTime;
+        this.millisInUtc =  utcMillis(instance);
+    }
+
+    @Override
+    public int getMillisOfSecond() {
+        return this.instance.get(ChronoField.MILLI_OF_SECOND);
+    }
+
+    @Override
+    public long getMilliseconds() {
+        return this.instance.toInstant().toEpochMilli();
+    }
+
+    @Override
+    public long getMillisecondsInUtc() {
+        return millisInUtc;
+    }
+
+    @Override
+    public String getString() {
+        return DateTimeUtil.jodaFormat(instance);
+    }
+
+    @Override
+    public int getTimeZoneOffsetHours() {
+        return this.instance.getOffset().getTotalSeconds()/ SECONDS_IN_HOUR;
+    }
+
+    @Override
+    public String getTimeZoneId() {
+        return this.instance.getZone().getId();
+    }
+
+    @Override
+    public Calendar toCalendar() {
+        return toCalendar(null);
+    }
+
+    @Override
+    public Calendar toCalendar( Locale locale ) { 
+        Calendar calendar = locale != null ? Calendar.getInstance(locale) : Calendar.getInstance();
+        calendar.setTimeInMillis(getMilliseconds());
+        return calendar;
+    }
+
+    @Override
+    public Date toDate() {
+        return new Date(getMilliseconds());
+    }
+
+    @Override
+    public ZonedDateTime toZonedDateTime() {
+        return instance;
+    }
+
+    @Override
+    public LocalDateTime toLocalDateTime() {
+        return instance.toLocalDateTime();
```

Comment:
You could move any of these methods that use `instance` to be default methods on the interface that use `toZonedDateTime()`. It's certainly not wrong to do it like you did, especially since there is no other impl, and because there's not much point in doing this _for the sake of moving to Java 8_. But as a general rule default methods are more idiomatic in Java 8, and it really does help to keep impls simple.